README.md
# SKLoader

Now you can animate any kind of loader with SKLoader framework.
Have a look at some examples below.

![earthloader](https://user-images.githubusercontent.com/40866243/43359582-21dc18c8-92c2-11e8-858e-d440fbddd935.gif)
![foodloader](https://user-images.githubusercontent.com/40866243/43359583-22090f54-92c2-11e8-9b40-a90426e33b7c.gif)
![pageloader](https://user-images.githubusercontent.com/40866243/43359584-22308e76-92c2-11e8-866e-840e6212b4f3.gif)
![simpleloader](https://user-images.githubusercontent.com/40866243/43359586-227ff1a0-92c2-11e8-8279-76bcd4c79d51.gif)
![uploadeloader](https://user-images.githubusercontent.com/40866243/43359587-22a7e19c-92c2-11e8-9c94-4bb21de3ff2c.gif)
![textloader](https://user-images.githubusercontent.com/40866243/43359611-a5174f1e-92c2-11e8-9ecc-34990aa571d9.gif)
![quoteloader](https://user-images.githubusercontent.com/40866243/43359585-2258e34e-92c2-11e8-948a-66d48f456fb7.gif)

## Requirements

`SKLoader` works for iOS 9.0+ and requires ARC to buid.
It requires gif to animate. Pre loaded with some gifs in example project. You can use your own gifs to animate in loader.

## Add SKLoader to your project

### Cocoapods ( Highly recommended way is to use cocoapods)
`pod 'SKLoader'`

## Usage:

1. import SKLoader
2. SKLoader works, as a singleton shared instance, or you can create an instance of SKLoader and use it.
3. Start Using It, by below sytax using shared instace. You can set up your SKLoader at didfinishlaunchingwithoptions method, or wherever you want to use. 

### Shared Instance:

```swift
SKLoader.sharedInstance.gifName = "SimpleLoader"
SKLoader.sharedInstance.animationSpeed = 0.01
SKLoader.sharedInstance.showLoader()
```
### SKLoader Instace:

```swift
let loader = SKLoader()
loader.gifName = "SimpleLoader"
loader.animationSpeed = 0.01
loader.showLoader()
Timer.scheduledTimer(withTimeInterval: 3.5, repeats: false) { (boolean) in
    loader.hideLoader()
}
```
### Public objects to finetune your animtaions:
```swift
gifName
backGroundColor
backgroundAlpha
animationSpeed
viewSize
viewContentMode
viewBackgroundColor
```
### Note:
1. Drag and drop a gif to your working 

### Noticeable mentions

[GitHub] (https://github.com/swiftgif/SwiftGif)

## License

This code is distributed under the terms and conditions of the [MIT license](LICENSE).

## Change-log

A brief summary of each release can be found in the [CHANGELOG](CHANGELOG.md). 


CHANGELOG.md
**Version 0.0.1** @ 04.08.2018

- Initial release.


SKLoader/SKLoader.swift
//
//  SKLoader.swift
//  SKLoader
//
//  Created by Srujan k on 23/07/18.
//  Copyright © 2018 Srujan k. All rights reserved.
//

import UIKit
public var animSpeed = 0.1
public class SKLoader: UIView {

    // MARK: - Public Variables
    public static let sharedInstance = SKLoader()
    public var gifName: String = ""
    public var backGroundColor: UIColor? = .black
    public var backgroundAlpha: CGFloat = 0.3
    public var animationSpeed = 0.1
    public var viewSize: CGSize = CGSize(width: 250, height: 150)
    public var viewContentMode : UIViewContentMode = .scaleAspectFit
    public var viewBackgroundColor: UIColor = UIColor.clear
    
    // MARK: - Local Variables
    private lazy var backgroundView: UIView = {
        let backgroundView = UIView(frame: UIScreen.main.bounds)
        backgroundView.backgroundColor = backGroundColor?.withAlphaComponent(backgroundAlpha)
        return backgroundView
    }()
    private lazy var gifImage: UIImageView = {
        let gifImage = UIImageView(frame: CGRect(x: 0, y: 0, width: viewSize.width, height: viewSize.height))
        gifImage.center = backgroundView.center
        gifImage.contentMode = viewContentMode
        gifImage.backgroundColor = viewBackgroundColor
        gifImage.loadGif(name: gifName)
        return gifImage
    }()
    
    // MARK: - Public Methods
    public func showLoader(){
        animSpeed = animationSpeed
        self.addSubview(backgroundView)
        self.backgroundView.addSubview(gifImage)
        self.backgroundView.bringSubview(toFront: self.gifImage)
        UIApplication.shared.keyWindow?.addSubview(self.backgroundView)
    }
    public func hideLoader(){
        self.backgroundView.removeFromSuperview()
    }
}


SKLoader/UIImage+Gif.swift
//
//  Gif.swift
//  SwiftGif
//
//  Created by Arne Bahlo on 07.06.14.
//  Copyright (c) 2014 Arne Bahlo. All rights reserved.
//

import UIKit
import ImageIO

extension UIImageView {

    public func loadGif(name: String) {
        DispatchQueue.global().async {
            let image = UIImage.gif(name: name)
            DispatchQueue.main.async {
                self.image = image
            }
        }
    }

    @available(iOS 9.0, *)
    public func loadGif(asset: String) {
        DispatchQueue.global().async {
            let image = UIImage.gif(asset: asset)
            DispatchQueue.main.async {
                self.image = image
            }
        }
    }

}

extension UIImage {

    public class func gif(data: Data) -> UIImage? {
        // Create source from data
        guard let source = CGImageSourceCreateWithData(data as CFData, nil) else {
            print("SwiftGif: Source for the image does not exist")
            return nil
        }

        return UIImage.animatedImageWithSource(source)
    }

    public class func gif(url: String) -> UIImage? {
        // Validate URL
        guard let bundleURL = URL(string: url) else {
            print("SwiftGif: This image named \"\(url)\" does not exist")
            return nil
        }

        // Validate data
        guard let imageData = try? Data(contentsOf: bundleURL) else {
            print("SwiftGif: Cannot turn image named \"\(url)\" into NSData")
            return nil
        }

        return gif(data: imageData)
    }

    public class func gif(name: String) -> UIImage? {
        // Check for existance of gif
        guard let bundleURL = Bundle.main
          .url(forResource: name, withExtension: "gif") else {
            print("SwiftGif: This image named \"\(name)\" does not exist")
            return nil
        }

        // Validate data
        guard let imageData = try? Data(contentsOf: bundleURL) else {
            print("SwiftGif: Cannot turn image named \"\(name)\" into NSData")
            return nil
        }

        return gif(data: imageData)
    }

    @available(iOS 9.0, *)
    public class func gif(asset: String) -> UIImage? {
        // Create source from assets catalog
        guard let dataAsset = NSDataAsset(name: asset) else {
            print("SwiftGif: Cannot turn image named \"\(asset)\" into NSDataAsset")
            return nil
        }

        return gif(data: dataAsset.data)
    }

    internal class func delayForImageAtIndex(_ index: Int, source: CGImageSource!) -> Double {
        var delay = 0.1

        // Get dictionaries
        let cfProperties = CGImageSourceCopyPropertiesAtIndex(source, index, nil)
        let gifPropertiesPointer = UnsafeMutablePointer<UnsafeRawPointer?>.allocate(capacity: 0)
        if CFDictionaryGetValueIfPresent(cfProperties, Unmanaged.passUnretained(kCGImagePropertyGIFDictionary).toOpaque(), gifPropertiesPointer) == false {
            return delay
        }

        let gifProperties:CFDictionary = unsafeBitCast(gifPropertiesPointer.pointee, to: CFDictionary.self)

        // Get delay time
        var delayObject: AnyObject = unsafeBitCast(
            CFDictionaryGetValue(gifProperties,
                Unmanaged.passUnretained(kCGImagePropertyGIFUnclampedDelayTime).toOpaque()),
            to: AnyObject.self)
        if delayObject.doubleValue == 0 {
            delayObject = unsafeBitCast(CFDictionaryGetValue(gifProperties,
                Unmanaged.passUnretained(kCGImagePropertyGIFDelayTime).toOpaque()), to: AnyObject.self)
        }

        delay = delayObject as? Double ?? 0

        if delay < animSpeed {
            delay = animSpeed // Make sure they're not too fast
        }

        return delay
    }

    internal class func gcdForPair(_ a: Int?, _ b: Int?) -> Int {
        var a = a
        var b = b
        // Check if one of them is nil
        if b == nil || a == nil {
            if b != nil {
                return b!
            } else if a != nil {
                return a!
            } else {
                return 0
            }
        }

        // Swap for modulo
        if a! < b! {
            let c = a
            a = b
            b = c
        }

        // Get greatest common divisor
        var rest: Int
        while true {
            rest = a! % b!

            if rest == 0 {
                return b! // Found it
            } else {
                a = b
                b = rest
            }
        }
    }

    internal class func gcdForArray(_ array: Array<Int>) -> Int {
        if array.isEmpty {
            return 1
        }

        var gcd = array[0]

        for val in array {
            gcd = UIImage.gcdForPair(val, gcd)
        }

        return gcd
    }

    internal class func animatedImageWithSource(_ source: CGImageSource) -> UIImage? {
        let count = CGImageSourceGetCount(source)
        var images = [CGImage]()
        var delays = [Int]()

        // Fill arrays
        for i in 0..<count {
            // Add image
            if let image = CGImageSourceCreateImageAtIndex(source, i, nil) {
                images.append(image)
            }

            // At it's delay in cs
            let delaySeconds = UIImage.delayForImageAtIndex(Int(i),
                source: source)
            delays.append(Int(delaySeconds * 1000.0)) // Seconds to ms
        }

        // Calculate full duration
        let duration: Int = {
            var sum = 0

            for val: Int in delays {
                sum += val
            }

            return sum
            }()

        // Get frames
        let gcd = gcdForArray(delays)
        var frames = [UIImage]()

        var frame: UIImage
        var frameCount: Int
        for i in 0..<count {
            frame = UIImage(cgImage: images[Int(i)])
            frameCount = Int(delays[Int(i)] / gcd)

            for _ in 0..<frameCount {
                frames.append(frame)
            }
        }

        // Heyhey
        let animation = UIImage.animatedImage(with: frames,
            duration: Double(duration) / 1000.0)

        return animation
    }

}


Example/Example/AppDelegate.swift
//
//  AppDelegate.swift
//  Example
//
//  Created by Srujan k on 23/07/18.
//  Copyright © 2018 Srujan k. All rights reserved.
//

import UIKit
import SKLoader

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?


    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
        setUpSKLoader()
        return true
    }
    func setUpSKLoader() {
        SKLoader.sharedInstance.gifName = "SimpleLoader"
        SKLoader.sharedInstance.animationSpeed = 0.01
    }
    func applicationWillResignActive(_ application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and invalidate graphics rendering callbacks. Games should use this method to pause the game.
    }

    func applicationDidEnterBackground(_ application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }

    func applicationWillEnterForeground(_ application: UIApplication) {
        // Called as part of the transition from the background to the active state; here you can undo many of the changes made on entering the background.
    }

    func applicationDidBecomeActive(_ application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }

    func applicationWillTerminate(_ application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }


}



SKLoaderTests/SKLoaderTests.swift
//
//  SKLoaderTests.swift
//  SKLoaderTests
//
//  Created by Srujan k on 23/07/18.
//  Copyright © 2018 Srujan k. All rights reserved.
//

import XCTest
@testable import SKLoader

class SKLoaderTests: XCTestCase {
    
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    func testExample() {
        // This is an example of a functional test case.
        // Use XCTAssert and related functions to verify your tests produce the correct results.
    }
    
    func testPerformanceExample() {
        // This is an example of a performance test case.
        self.measure {
            // Put the code you want to measure the time of here.
        }
    }
    
}


Example/Example/ExampleViewController.swift
//
//  ExampleViewController.swift
//  Example
//
//  Created by Srujan k on 23/07/18.
//  Copyright © 2018 Srujan k. All rights reserved.
//

import UIKit
import SKLoader

class ExampleViewController: UITableViewController {
    
    // MARK: - Controller Cycle
    override func viewDidLoad() {
        super.viewDidLoad()
    }
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
        
    // MARK: - Table view data source
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 8
    }
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath){
        switch indexPath.row {
        case 0:
            let loader = SKLoader()
            loader.gifName = "SimpleLoader"
            loader.animationSpeed = 0.01
            loader.showLoader()
            Timer.scheduledTimer(withTimeInterval: 3.0, repeats: false) { (boolean) in
                loader.hideLoader()
            }

            break
        case 1:
            let loader = SKLoader()
            loader.gifName = "Hello"
            loader.animationSpeed = 0.01
            loader.viewSize = CGSize(width: 250, height: 120)
            loader.showLoader()
            Timer.scheduledTimer(withTimeInterval: 3.0, repeats: false) { (boolean) in
                loader.hideLoader()
            }
            break
        case 2:
            let loader = SKLoader()
            loader.gifName = "earth"
            loader.animationSpeed = 0.01
            loader.viewContentMode = .scaleAspectFit
            loader.showLoader()
            Timer.scheduledTimer(withTimeInterval: 3.0, repeats: false) { (boolean) in
                loader.hideLoader()
            }
            break
        case 3:
            let loader = SKLoader()
            loader.gifName = "PageLoading"
            loader.animationSpeed = 0.01
            loader.viewBackgroundColor = UIColor.white
            loader.viewSize = CGSize(width: 200, height: 150)
            loader.showLoader()
            Timer.scheduledTimer(withTimeInterval: 3.0, repeats: false) { (boolean) in
                loader.hideLoader()
            }
            break
        case 4:
            let loader = SKLoader()
            loader.gifName = "Upload"
            loader.animationSpeed = 0.01
            loader.showLoader()
            Timer.scheduledTimer(withTimeInterval: 3.5, repeats: false) { (boolean) in
                loader.hideLoader()
            }
            break
        case 5:
            let loader = SKLoader()
            loader.gifName = "Quote"
            loader.animationSpeed = 0.01
            loader.viewSize = CGSize(width: UIScreen.main.bounds.width - 20.0, height: 150)
            loader.showLoader()
            Timer.scheduledTimer(withTimeInterval: 3.0, repeats: false) { (boolean) in
                loader.hideLoader()
            }
            break
        case 6:
            let loader = SKLoader()
            loader.gifName = "FoodLoading"
            loader.animationSpeed = 0.01
            loader.showLoader()
            Timer.scheduledTimer(withTimeInterval: 3.0, repeats: false) { (boolean) in
                loader.hideLoader()
            }
            break
        case 7:
           SKLoader.sharedInstance.showLoader()
            Timer.scheduledTimer(withTimeInterval: 3.0, repeats: false) { (boolean) in
                SKLoader.sharedInstance.hideLoader()
            }
            break
        default:
            break
        }

    }
    // MARK: - Action Methods
    @objc func closeLoader(){
        SKLoader.sharedInstance.hideLoader()
    }
}
